File: packages/bonsoir_android/android/src/main/kotlin/fr/skyost/bonsoir/broadcast/BonsoirServiceBroadcast.kt
package fr.skyost.bonsoir.broadcast

import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import fr.skyost.bonsoir.BonsoirAction
import fr.skyost.bonsoir.BonsoirService
import fr.skyost.bonsoir.Generated
import io.flutter.plugin.common.BinaryMessenger

/**
 * Allows to broadcast a NSD service on local network.
 *
 * @param id The listener identifier.
 * @param printLogs Whether to print debug logs.
 * @param onDispose Triggered when this instance is being disposed.
 * @param nsdManager The NSD manager.
 * @param messenger The Flutter binary messenger.
 * @param service The Bonsoir service to broadcast.
 */
class BonsoirServiceBroadcast(
    id: Int,
    printLogs: Boolean,
    onDispose: Runnable,
    nsdManager: NsdManager,
    messenger: BinaryMessenger,
    private val service: BonsoirService
) : BonsoirAction(
    id,
    "broadcast",
    Generated.broadcastMessages,
    printLogs,
    onDispose,
    nsdManager,
    messenger
), NsdManager.RegistrationListener {

    /**
     * Starts the service registration.
     */
    fun start() {
        if (!isActive) {
            nsdManager.registerService(service.toNsdService(), NsdManager.PROTOCOL_DNS_SD, this)
        }
    }

    override fun onServiceRegistered(service: NsdServiceInfo) {
        makeActive()
        if (this.service.name != service.serviceName) {
            val oldName = this.service.name
            this.service.name = service.serviceName
            onSuccess(Generated.broadcastNameAlreadyExists, this.service, parameters = listOf(oldName))
        }
        onSuccess(Generated.broadcastStarted, this.service)
    }

    override fun onRegistrationFailed(service: NsdServiceInfo, errorCode: Int) {
        onError(parameters = listOf(this.service, errorCode), details = errorCode)
        dispose()
    }

    override fun onServiceUnregistered(service: NsdServiceInfo) {
        val wasActive = isActive
        makeUnactive()
        onSuccess(Generated.broadcastStopped, this.service)
        dispose(wasActive)
    }

    override fun onUnregistrationFailed(service: NsdServiceInfo, errorCode: Int) {
        onError("Bonsoir service unregistration failed : %s (error : %s).", listOf(this.service, errorCode), errorCode)
    }

    override fun stop() {
        nsdManager.unregisterService(this)
    }
}

File: packages/bonsoir_darwin/darwin/bonsoir_darwin/Sources/bonsoir_darwin/Broadcast/BonsoirServiceBroadcast.swift
#if canImport(Flutter)
    import Flutter
#endif
#if canImport(FlutterMacOS)
    import FlutterMacOS
#endif
import Network

/// Allows to broadcast a given service to the local network.
@available(iOS 13.0, macOS 10.15, *)
class BonsoirServiceBroadcast: BonsoirAction {
    /// The advertised service.
    private let service: BonsoirService

    /// The reference to the registering..
    private var sdRef: DNSServiceRef?

    /// Initializes this class.
    public init(id: Int, printLogs: Bool, onDispose: @escaping () -> Void, messenger: FlutterBinaryMessenger, service: BonsoirService) {
        self.service = service
        super.init(id: id, action: "broadcast", logMessages: Generated.broadcastMessages, printLogs: printLogs, onDispose: onDispose, messenger: messenger)
    }

    /// Starts the broadcast.
    public func start() {
        var txtRecord = TXTRecordRef()
        TXTRecordCreate(&txtRecord, 0, nil)
        for (key, value) in service.attributes {
            guard let valueData = value.data(using: .utf8) else { continue }
            TXTRecordSetValue(&txtRecord, key, UInt8(valueData.count), [UInt8](valueData))
        }
        let error = DNSServiceRegister(&sdRef, 0, 0, service.name, service.type, "local.", service.host, CFSwapInt16HostToBig(UInt16(service.port)), TXTRecordGetLength(&txtRecord), TXTRecordGetBytesPtr(&txtRecord), BonsoirServiceBroadcast.registerCallback as DNSServiceRegisterReply, Unmanaged.passUnretained(self).toOpaque())
        if error == kDNSServiceErr_NoError {
            log(logMessages[Generated.broadcastInitialized]!, [service])
            DNSServiceProcessResult(sdRef)
        } else {
            onError(parameters: [service, error], details: error)
            dispose()
        }
    }

    override public func dispose() {
        DNSServiceRefDeallocate(sdRef)
        onSuccess(eventId: Generated.broadcastStopped, service: service)
        super.dispose()
    }

    /// Callback triggered by`DNSServiceRegister`.
    private static let registerCallback: DNSServiceRegisterReply = { _, _, errorCode, name, _, _, context in
        let broadcast = Unmanaged<BonsoirServiceBroadcast>.fromOpaque(context!).takeUnretainedValue()
        if errorCode == kDNSServiceErr_NoError {
            let newName = name == nil ? nil : String(cString: name!)
            if newName != nil && broadcast.service.name != newName {
                let oldName = broadcast.service.name
                broadcast.service.name = newName!
                broadcast.onSuccess(eventId: Generated.broadcastNameAlreadyExists, service: broadcast.service, parameters: [oldName])
            }
            broadcast.onSuccess(eventId: Generated.broadcastStarted, service: broadcast.service)
        } else {
            broadcast.onError(parameters: [broadcast.service, errorCode], details: errorCode)
            broadcast.dispose()
        }
    }
}
